fairlearn/linear_model/_constrained_logistic.py

```diff
@@ -0,0 +1,813 @@
+        columns to filter out, either as strings (DataFrame) or numbers (numpy)
+
+    Returns
+    -------
+    X : numpy.ndarray or pandas.DataFrame
+        Feature data with one-hot-encoded values
+
+    renamed_sensitive_feature_ids : list
+        The renamed sensitive feature ids, either as strings or as numbers
+
+    categories : list
+        The categories from the encoder
+    """
+    enc = OneHotEncoder(handle_unknown="ignore")
+    if isinstance(X, pd.DataFrame):
+        transformed = enc.fit_transform(X[sensitive_feature_ids]).toarray()
+        renamed_sensitive_feature_ids = list(enc.get_feature_names_out())
+        # Create a Pandas DataFrame of the hot encoded column
+        ohe_df = pd.DataFrame(
+            transformed, columns=renamed_sensitive_feature_ids
+        )
+        # concat with original data, drop the original sensitive_feature_ids
+        X = pd.concat([X, ohe_df], axis=1).drop(sensitive_feature_ids, axis=1)
+    else:  # Numpy array
+        transformed = enc.fit_transform(X[:, sensitive_feature_ids]).toarray()
+        # Delete the old column and append the transformed columns
+        X_without_sensitive = np.delete(X, sensitive_feature_ids, axis=1)
+        X = np.append(X_without_sensitive, transformed, axis=1)
+        # Need to return the new transformed sensitive feature ids since there are more columns now
+        renamed_sensitive_feature_ids = list(
+            range(X_without_sensitive.shape[1], X.shape[1])
+        )
+    categories = enc.categories_
+
+    return X, renamed_sensitive_feature_ids, categories
+
+
+class ConstrainedLogisticRegression(LogisticRegression):
+    r"""Constrained logistic regression.
+
+    This implementation closely follows the implementation in
+    :class:`sklearn.linear_model.LogisticRegression`, and is constrained
+    via the same way as in
+    `Zafar et al. (2017) <https://proceedings.mlr.press/v54/zafar17a.html>`_ [1]_.
+
+    The constraint is defined as the covariance between the sensitive
+    features :math:`\left\{\mathbf{z}_{i}\right\}_{i=1}^{N}`, and the
+    signed distance from the users` feature vectors to the decision
+    boundary :math:`\left\{d_{\boldsymbol{\theta}}\left(\mathbf{x}_{i}\right)\right\}_{i=1}^{N}`.
+    This is mathemetically described as follows:
+
+    .. math::
+      \operatorname{Cov}\left(\mathbf{z}, d_{\boldsymbol{\theta}}(\mathbf{x})\right)
+      =\mathbb{E}\left[(\mathbf{z}-\overline{\mathbf{z}})
+        d_{\boldsymbol{\theta}}(\mathbf{x})\right]
+        -\mathbb{E}[(\mathbf{z}-\overline{\mathbf{z}})]
+        \bar{d}_{\boldsymbol{\theta}}(\mathbf{x})\\
+      \approx \frac{1}{N} \sum_{i=1}^{N}\left(\mathbf{z}_{i}-\overline{\mathbf{z}}\right)
+      d_{\boldsymbol{\theta}}\left(\mathbf{x}_{i}\right)
+
+    The reasoning behind this derivation is further explained in [1]_.
+
+    Parameters
+    ----------
+    constraints : {'demographic_parity', 'none'}, default='demographic_parity'
```

Comment:
I agree with your point, but am also unsure what would be a good name replacement. Would something like _demographic_parity_covariance_ work, or would that only cause more confusion?

Reply:
@fairlearn/fairlearn-maintainers WDYT?

Reply:
I agree that the name should ideally be a different one. I'm agnostic on which we choose:
- demographic_parity_covariance or covariance_demographic_parity
- something based on the names of the authors of the paper
- convex_demographic_parity (don't really like this one)

Reply:
If we do not mind long strings, my preference would probably be *demographic_parity_covariance*. I'm personally not a huge fan of naming things after authors because it's harder to remember what's what :)